### PR TITLE
Add dart define from file

### LINKED ISSUE #1742: Patrol CLI doesn't support `--dart-define-from-file`
See discussion in #1477, especially [this comment](https://github.com/leancodepl/patrol/pull/1477#issuecomment-1635860323).

If we want to be consistent with Flutter, we should wait for https://github.com/flutter/flutter/issues/130599 to be resolved.

### PR DESCRIPTION
**Pull request by pdenert**
Close #1742

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -153,6 +153,21 @@ abstract class NativeAutomator<IOSServer, AndroidServer, DartClient> {
...
+If you have previously activated patrol_cli run:
+
+```bash
+dart pub global deactivate patrol_cli
+```
+
+then
+
+```bash
+cd patrol/packages/patrol_cli
```
Comment (mateuszwojtczak) [inline README.md:167]: I'd say `cd packages/patrol_cli`, because it's best to assume that the working directory would be the README's one

File: packages/patrol_cli/lib/src/commands/build_android.dart [hunk 2]
```diff
@@ -121,12 +122,31 @@ class BuildAndroidCommand extends PatrolCommand {
...
+        'Received path for --dart-define-from-file: $dartDefineFromFilePath',
+      );
+      dartDefineFromFiles = mergeKeys(
+        json: dartDefineFromFiles,
+        dartDefines:
+            _dartDefinesReader.fromConfigFile(path: dartDefineFromFilePath),
+      );
+    }
+
+    final dartDefinesMerged =
```
Comment (mateuszwojtczak) [inline packages/patrol_cli/lib/src/commands/build_android.dart:140]: `mergedDartDefines` sounds more English

File: packages/patrol_cli/lib/src/commands/test.dart [hunk 3]
```diff
@@ -179,12 +180,31 @@ See https://github.com/leancodepl/patrol/issues/1316 to learn more.
       );
     }
 
+    final dartDefineFromFilePaths = stringsArg('dart-define-from-file');
```
Comment (mateuszwojtczak) [inline packages/patrol_cli/lib/src/commands/test.dart:183]: L183-200 are pretty much the same as L125-142 in `build_android.dart` and L192-209 in `develop.dart`.

I think it's the same behavior and we could extract it and have in one place, WDYT?

Comment (mateuszwojtczak) [reply packages/patrol_cli/lib/src/commands/test.dart:183]: in the end you just want some function that returns `mergedDartDefines` and `dartDefineFromFilePaths`

File: packages/patrol_cli/lib/src/crossplatform/app_options.dart [hunk 4]
```diff
@@ -305,3 +319,17 @@ class MacOSAppOptions {
     return cmd;
   }
 }
+
+Map<String, String> mergeKeys({
```
Comment (mateuszwojtczak) [inline packages/patrol_cli/lib/src/crossplatform/app_options.dart:323]: why is this here? is it related in any way to `AppOptions`?

File: packages/patrol_cli/lib/src/dart_defines_reader.dart [hunk 5]
```diff
@@ -25,6 +27,19 @@ class DartDefinesReader {
     return _parse(lines);
   }
 
+  Map<String, dynamic> fromConfigFile({required String path}) {
+    final filePath = _fs.path.join(_projectRoot.path, path);
+    final file = _fs.file(filePath);
+
+    if (!file.existsSync()) {
```
Comment (mateuszwojtczak) [inline packages/patrol_cli/lib/src/dart_defines_reader.dart:34]: I think this is a very controversial decision. I can imagine this situation: you want to inject dart defines from a file and you make a typo in the path, so the file doesn't exist and your tests are being run but it just behaves in a different way than you expected.

This kind of behaviors can lead to hours-long debugging that gives you a headache but you just made a typo in a path.

BUT if we just gave a warning to the user, that this file doesn't exist, I would be so grateful as a user :D

File: packages/patrol_cli/lib/src/dart_defines_reader.dart [hunk 6]
```diff
@@ -25,6 +27,19 @@ class DartDefinesReader {
...
+      return {};
+    }
+
+    final jsonString = file.readAsStringSync();
+    final json = jsonDecode(jsonString) as Map<String, dynamic>;
+    return json;
```
Comment (mateuszwojtczak) [inline packages/patrol_cli/lib/src/dart_defines_reader.dart:40]: we can just return in L39

**Review by mateuszwojtczak — CHANGES_REQUESTED** [review]

File: packages/patrol_cli/lib/src/dart_defines_reader.dart [hunk 7]
```diff
@@ -25,6 +27,19 @@ class DartDefinesReader {
...
+  }
+
```
Comment (nikitadol) [inline packages/patrol_cli/lib/src/dart_defines_reader.dart:42]: If the decision was made to completely parse the files into separate `define`'s, then I would like it to work the same way as in `flutter_tools`:
- fail if file is missing
- do not add `project Root.path` at beginning of file path
- support more than just `JSON` (maybe someone is using this)

How this is implemented in `flutter_tools`:
https://github.com/flutter/flutter/blob/64dd1ca9bc32fbc767a3f63e4c9c5ae2fc355380/packages/flutter_tools/lib/src/runner/flutter_command.dart#L1452-L1494

File: packages/patrol_cli/lib/src/dart_defines_reader.dart (hunk 5 above)
Comment (pdenert) [reply packages/patrol_cli/lib/src/dart_defines_reader.dart:38]: Ok, I've add logger warning

File: packages/patrol_cli/lib/src/commands/test.dart (hunk 3 above)
Comment (pdenert) [reply packages/patrol_cli/lib/src/commands/test.dart:183]: I've extracted this common code to new function

File: dev/e2e_app/integration_test/define_test.dart
Comment (piotruela) [inline dev/e2e_app/integration_test/define_test.dart:1]: Won't this test cause all workflows with patrol tests in our repository to fail? I think we should exclude this test from all of them

File: packages/patrol_cli/lib/src/dart_defines_reader.dart (hunk 7 above)
Comment (pdenert) [reply packages/patrol_cli/lib/src/dart_defines_reader.dart]: Thanks. Good notice! I reused flutter_tools code for parsing. You can check it out now

**Review by mateuszwojtczak — CHANGES_REQUESTED** [review]
Great, one last thing - if we allow for .env style files, let's have at least one test about it

**Comment by pdenert** [comment]
> Great, one last thing - if we allow for .env style files, let's have at least one test about it 

Done @mateuszwojtczak

File: dev/e2e_app/integration_test/define_test.dart
Comment (pdenert) [reply dev/e2e_app/integration_test/define_test.dart:1]: You're right. Need to manage that

**Review by mateuszwojtczak — APPROVED** [review]
LGTM

**Review by piotruela — APPROVED** [review]

File: packages/patrol_cli/lib/src/commands/dart_define_utils.dart [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+/// Merge two map with overriding values for duplicated keys, by [dartDefines] values.
+Map<String, String> mergeKeys({
+  required Map<String, dynamic> json,
+  required Map<String, dynamic> dartDefines,
+}) {
+  final modified = Map<String, String>.from(json);
+
+  if (dartDefines.isNotEmpty) {
+    dartDefines.forEach((key, value) {
+      modified[key] = value as String;
```
Comment (nikitadol) [inline packages/patrol_cli/lib/src/commands/dart_define_utils.dart:12]: @pdenert 
I tried this change in our project and realized that not everything is in the form of strings, so the code crashes here - I suspect that other projects may have a similar problem

Comment (pdenert) [reply packages/patrol_cli/lib/src/commands/dart_define_utils.dart:12]: You're right. Can you create issue for that? We will try to address this ASAP

Comment (nikitadol) [reply packages/patrol_cli/lib/src/commands/dart_define_utils.dart:12]: Done https://github.com/leancodepl/patrol/issues/2242
